File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/say/say.go
package say

import "strings"

var ones = []string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
var teens = []string{"ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"}
var tens = []string{"", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"}
var scales = []string{"", "thousand", "million", "billion", "trillion"}

func Say(n int64) (string, bool) {
	if n < 0 || n > 999999999999 {
		return "", false
	}
	if n == 0 {
		return "zero", true
	}

	chunks := splitIntoChunks(n)
	reversedChunks := reverseChunks(chunks)
	var parts []string

	for i, chunk := range reversedChunks {
		if chunk == 0 {
			continue
		}
		part := numToWordsUnder1000(chunk)
		scaleIndex := len(reversedChunks) - 1 - i
		if scaleIndex >= len(scales) {
			return "", false
		}
		scale := scales[scaleIndex]
		if scale != "" {
			part += " " + scale
		}
		parts = append(parts, part)
	}

	return strings.Join(parts, " "), true
}

func splitIntoChunks(n int64) []int64 {
	var chunks []int64
	for n > 0 {
		chunks = append(chunks, n%1000)
		n /= 1000
	}
	return chunks
}

func reverseChunks(chunks []int64) []int64 {
	reversed := make([]int64, len(chunks))
	for i, j := 0, len(chunks)-1; j >= 0; i, j = i+1, j-1 {
		reversed[i] = chunks[j]
	}
	return reversed
}

func numToWordsUnder1000(n int64) string {
	var parts []string
	hundreds := n / 100
	remainder := n % 100

	if hundreds > 0 {
		parts = append(parts, ones[hundreds]+" hundred")
	}
	if remainder > 0 {
		parts = append(parts, numToWordsUnder100(remainder))
	}
	return strings.Join(parts, " ")
}

func numToWordsUnder100(n int64) string {
	if n < 10 {
		return ones[n]
	}
	if n < 20 {
		return teens[n-10]
	}
	tensDigit := n / 10
	onesDigit := n % 10
	tensPart := tens[tensDigit]
	if onesDigit == 0 {
		return tensPart
	}
	return tensPart + "-" + ones[onesDigit]
}

File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/say/say.js
//
// This is only a SKELETON file for the 'Say' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const say = (n) => {
  if (typeof n !== 'number' || n < 0 || n > 999999999999 || !Number.isInteger(n)) {
    throw new Error('Number must be between 0 and 999,999,999,999.');
  }

  if (n === 0) {
    return 'zero';
  }

  const ones = [
    'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine',
    'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen',
    'seventeen', 'eighteen', 'nineteen'
  ];
  
  const tens = [
    '', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety'
  ];

  const convertLessThanOneHundred = (number) => {
    if (number < 20) {
      return ones[number];
    }
    const tenPart = tens[Math.floor(number / 10)];
    const remainder = number % 10;
    if (remainder === 0) {
      return tenPart;
    }
    return tenPart + '-' + ones[remainder];
  };

  const convertLessThanOneThousand = (number) => {
    if (number < 100) {
      return convertLessThanOneHundred(number);
    }
    const hundredPart = ones[Math.floor(number / 100)] + ' hundred';
    const remainder = number % 100;
    if (remainder === 0) {
      return hundredPart;
    }
    return hundredPart + ' ' + convertLessThanOneHundred(remainder);
  };

  // Split into chunks of thousands
  const chunks = [];
  let num = n;
  while (num > 0) {
    chunks.push(num % 1000);
    num = Math.floor(num / 1000);
  }
  chunks.reverse();

  const scales = ['', 'thousand', 'million', 'billion'];
  const parts = [];
  
  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    if (chunk === 0) continue; // Skip zero chunks
    
    const chunkText = convertLessThanOneThousand(chunk);
    const scaleIndex = chunks.length - 1 - i;
    const scale = scales[scaleIndex];
    
    parts.push(chunkText + (scale ? ' ' + scale : ''));
  }

  return parts.join(' ');
};
